### PR TITLE
[CDU] Various CostIndex Fixes

### LINKED ISSUE #2499: [BUG] Cost Index 0 ignored by CDU
**Mod Version**

latest master

**Describe the bug**

Entering a Cost Index of 0 into the CDU is ignored by the FMC, the field stays blank

**To Reproduce**

1. try entering a cost index of 0
2.
3.

**Expected behavior**

Cost Index 0 should give Maximum Range and be accepted by the FMC

**Actual behavior**

It ignors it

**References**

https://ansperformance.eu/library/airbus-cost-index.pdf

**Additional context**

Was this working before/when did the issue start occurring?
Was always wrong
Is this a problem in the vanilla unmodded game?
Yes

Discord username (if different from GitHub):

ChruutvoLuzi | SD-PB.ch#8902

### PR DESCRIPTION
**Pull request by ChruutvoLuzi**
Fixes #2499

## Summary of Changes

Changes Some Error Messages:

- [X] If CI < 0 = "FORMAT ERROR" (Previously NOT ALLOWED)
- [X] If CI ≠ Number = "FORMAT ERROR" (Previously NOT ALLOWED

**If someone wants to help out with an CostIndex Equation for the speeds, feel free to do so, as i am not the best in Math**

## Screenshots (if necessary)

https://cdn.discordapp.com/attachments/747622836381810739/789569394136317962/Microsoft_Flight_Simulator_2020.12.18_-_20.04.40.04.mp4

## References

102613176-b4133b80-413a-11eb-9b87-7015c5690315
efsdfsdf

## Additional context

This is my first try at coding something so it probably wont be perfect :)

Discord username (if different from GitHub):
ChruutvoLuzi | SD-PB.ch#8902

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by Watsi01** [comment]
Check out this. It's old but is has some good informations on CI (the later posts):

https://github.com/flybywiresim/a32nx/pull/1389

**Comment by ChruutvoLuzi** [comment]
> Check out this. It's old but is has some good informations on CI (the later posts):
> 
> 
> 
> https://github.com/flybywiresim/a32nx/pull/1389
> 
> 

Thanks! Stuff looks pretty complicated, not sure if i will be able to do anything usefull in that regard. Are you still working on it? If yes, I'll concentrate on other things then.

**Comment by Watsi01** [comment]
@ChruutvoLuzi No. I don't work on it. It exceeded my competence.

**Comment by donstim** [comment]
Thanks for picking this up. Do you have a reference for the "FORMAT ERROR" scratchpad message for CI <0?  I think if a number is input, but it is not in the allowed range of 0-999, the scratchpad error message should be "ENTRY OUT OF RANGE."
Screenshot (451)

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js [hunk 1]
```diff
@@ -547,9 +547,12 @@ class FMCMainDisplay extends BaseAirliners {
                     this.showErrorMessage("ENTRY OUT OF RANGE");
                     return false;
                 }
+            } else {
+                this.showErrorMessage("FORMAT ERROR");
+                return false;
             }
         }
-        this.showErrorMessage(this.defaultInputErrorMessage);
+        this.showErrorMessage("FORMAT ERROR");
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:555]: Maybe we should add those FORMAT ERROR strings as a class field like `this.defaultInputErrorMessage`.

**Comment by ChruutvoLuzi** [comment]
> Thanks for picking this up. Do you have a reference for the "FORMAT ERROR" scratchpad message for CI <0?  I think if a number is input, but it is not in the allowed range of 0-999, the scratchpad error message should be "ENTRY OUT OF RANGE."
> 
> Screenshot (451)
> 
> 

Not really a reference but only have this:
image

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js (hunk 1 above)
Comment (ChruutvoLuzi) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:555]: Sure! Will give it a go when i get some time

**Comment by donstim** [comment]
While I think it would be consistent with an out range error as described in the Honeywell document, I don't have any definitive reference source either, so unless you can get a pilot to try it out in the real airplane, might as well go with @beheh

**Comment by beheh** [comment]
Thanks for the mention I've since revised my theory and I think it's defined more clearly as I put in Discord. I think the two error error messages:

image

image


...correspond to these columns 2 and 3 here ("Format" and "Range" respectively):

image

My theory is that if column 2 is violated you get `FORMAT ERROR`, and if column 3 is violated you get `ENTRY OUT OF RANGE`.

This makes sense for flight level:

image

...where the values `ABC` and `1500` lead to `FORMAT ERROR`, but `FL460` or `46000` to `ENTRY OUT OF RANGE`.

For COST INDEX above, my theory is that it's impossible to get `ENTRY OUT OF RANGE`, because all formats (000-999) are in range, however, if you enter `-1`, `1000` or `ABC` you're violating `NNN` and therefore should receive a `FORMAT ERROR`.

This is why I believe the `FORMAT ERROR` is correct here. But obviously that could be wrong, it is just my best guess. Maybe @Pushok154 knows or could test this?

**Comment by MisterChocker** [comment]
To add to the consensus, I believe we should make an message object and predefine every possible message.
Thoughts?

**Comment by ChruutvoLuzi** [comment]
> To add to the consensus, I believe we should make a message object or dictionary and predefine every possible message.
> 
> Thoughts?

@MisterChocker sounds better yeah, i could give it a go if thats ok

**Comment by MisterChocker** [comment]
> @MisterChocker sounds better yeah, i could give it a go if thats ok

Already at it xD

**Comment by ChruutvoLuzi** [comment]
> 
> 
> > @MisterChocker sounds better yeah, i could give it a go if thats ok
> 
> Already at it xD

@MisterChocker oh, lol
aight gonna close this PR then as the CI is too complex for me and you are already doing the message stuff xF

**Comment by MisterChocker** [comment]
@ChruutvoLuzi can you write something in #fms channel on our discord? I'd like to share my thoughts.

**Pull request closed without merging**